src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,168 +23,324 @@ using namespace opossum::expression_functional;  // NOLINT
-      return LQPVisitation::VisitInputs;
-    }
+void recursive_gather_required_expressions(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, size_t>& outputs_visited_by_node) {
+  const auto additional_columns = gather_required_expressions(node);
+  auto& required_expressions = required_expressions_by_node[node];
+  required_expressions.insert(additional_columns.begin(), additional_columns.end());
+
+  ++outputs_visited_by_node[node];
+  if (outputs_visited_by_node[node] < node->output_count()) return;
 
-    auto pruned_column_ids = std::vector<ColumnID>{};
-    for (const auto& expression : node->column_expressions()) {
-      if (referenced_columns.find(expression) != referenced_columns.end()) {
-        continue;
+  // Once all nodes that may require columns from this node (i.e., this node's outputs) have been visited, we can
+  // recurse into this node's inputs.
+  for (const auto& input : {node->left_input(), node->right_input()}) {
+    if (!input) continue;
+
+    // Make sure the entry exists, then insert all expressions that the current node needs
+    required_expressions_by_node[input];
+    for (const auto& required_expression : required_expressions) {
+      // Add the columns needed here (and above) if they come from the input node
+      if (input->find_column_id(*required_expression)) {
+        required_expressions_by_node[input].emplace(required_expression);
       }
-      const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
-      pruned_column_ids.emplace_back(column_expression->column_reference.original_column_id());
     }
 
-    // We cannot create a Table without columns - since Chunks rely on their first column to determine their row count
-    if (pruned_column_ids.size() == node->column_expressions().size() && !pruned_column_ids.empty()) {
-      pruned_column_ids.pop_back();
+    recursive_gather_required_expressions(input, required_expressions_by_node, outputs_visited_by_node);
+  }
+}
+
+void prune_join_node(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node) {
+  // Sometimes, joins are not actually used to combine tables but only to check the existence of a tuple in a second
+  // table. Example: SELECT c_name FROM customer, nation WHERE c_nationkey = n_nationkey AND n_name = 'GERMANY'
+  // If the join is on a unique/primary key column, we can rewrite these joins into semi joins. If, however, the
+  // uniqueness is not guaranteed, we cannot perform the rewrite as non-unique joins could possibly emit a matching
+  // line more than once.
+
+  auto join_node = std::dynamic_pointer_cast<JoinNode>(node);
+  if (join_node->join_mode != JoinMode::Inner) return;
```

Comment:
Why do you check for inner? I see that this check is required for the semi optimization, but also for the column pruning?

Reply:
Ah ok, there is no column pruning taking place, right? The only thing we do in this method is the semi-join optimization. Was expecting something else from the name.

Apart from this optimization: could there by a case we actual column pruning could make sense in a join?

Reply:
> Ah ok, there is no column pruning taking place, right? The only thing we do in this method is the semi-join optimization. Was expecting something else from the name.

Changed to `try_join_to_semi_rewrite`.

> Apart from this optimization: could there by a case we actual column pruning could make sense in a join?

Absolutely, but given that our join operators currently do not support pruning, there is nothing we can do. Adding it to this rule is trivial.